From 40ebd51e15c3746206498ccc1edc8ba34541ff45 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Wed, 16 May 2018 21:48:50 +0100
Subject: [PATCH] backport PlayerInventory#getItemInMainHand


diff --git a/src/main/java/org/bukkit/inventory/PlayerInventory.java b/src/main/java/org/bukkit/inventory/PlayerInventory.java
index b79a4f02..75a9e380 100644
--- a/src/main/java/org/bukkit/inventory/PlayerInventory.java
+++ b/src/main/java/org/bukkit/inventory/PlayerInventory.java
@@ -112,6 +112,15 @@ public interface PlayerInventory extends Inventory {
      */
     public ItemStack getItemInHand();
 
+    // Alphheim Start - Backport getItemInMainHand
+    /**
+     * Returns the ItemStack currently hold
+     *
+     * @return The currently held ItemStack
+     */
+    public ItemStack getItemInMainHand();
+    // Alphheim end - Backport getItemInMainHand
+
     /**
      * Sets the item in hand
      *
-- 
2.17.0

