From 36de03e5bc4ef54718e106116582b92b363f7bc7 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sat, 6 Jan 2018 14:44:38 +0000
Subject: [PATCH] Don't tick 'unlimited' potion effects


diff --git a/src/main/java/net/minecraft/server/MobEffect.java b/src/main/java/net/minecraft/server/MobEffect.java
index c9f433f83..64c85dd55 100644
--- a/src/main/java/net/minecraft/server/MobEffect.java
+++ b/src/main/java/net/minecraft/server/MobEffect.java
@@ -91,7 +91,7 @@ public class MobEffect {
     }
 
     private int i() {
-        return --this.duration;
+        return duration == Integer.MAX_VALUE ? Integer.MAX_VALUE : --this.duration; // Alphheim - don't reduce the duration of 'unlimited' effects
     }
 
     public void b(EntityLiving entityliving) {
-- 
2.16.2

