From 464b4b057dd653e992f1369665a362eee07c579c Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Mon, 9 Apr 2018 08:21:27 +0100
Subject: [PATCH] Temp make PCM public for plugin being stupid


diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index 38586aad7..82130f514 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -349,7 +349,7 @@ public class PlayerChunkMap {
         return i * 16 - 16;
     }
 
-    class PlayerChunk {
+    public class PlayerChunk {
 
         private final List<EntityPlayer> b = Lists.newArrayList();
         private final ChunkCoordIntPair location;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index bb563b5dc..7c7011033 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -195,6 +195,9 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
     Builder<String, Object> serialize(Builder<String, Object> builder) {
         super.serialize(builder);
         if (hasOwner()) {
+
+
+
             return builder.put(SKULL_OWNER.BUKKIT, this.profile.getName());
         }
         return builder;
-- 
2.17.0

