From f99539ba8e745d51b51a07eec4b45acad3ce498e Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Thu, 26 Apr 2018 11:51:10 +0100
Subject: [PATCH] Prevent sign update NPEs due to horrible mods


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index c637c2bd..29a5dc69 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1851,7 +1851,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             String[] lines = new String[4];
 
             for (int i = 0; i < aichatbasecomponent.length; ++i) {
-                lines[i] = EnumChatFormat.a(aichatbasecomponent[i].c());
+                lines[i] = EnumChatFormat.a(aichatbasecomponent[i] != null ? aichatbasecomponent[i].c() : ""); // Alphheim - prevent NPEs due to horrible mods...
             }
             SignChangeEvent event = new SignChangeEvent((org.bukkit.craftbukkit.block.CraftBlock) player.getWorld().getBlockAt(x, y, z), this.server.getPlayer(this.player), lines);
             this.server.getPluginManager().callEvent(event);
-- 
2.17.0

