From 39f1bd85605bc60f6aca0d10f2d8bee7716f47e2 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sat, 19 May 2018 01:38:55 +0100
Subject: [PATCH] Add EffectRemovalNotification


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index c8b4e292..ab5d0e61 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -24,9 +24,12 @@ import org.bukkit.event.vehicle.VehicleExitEvent;
 // CraftBukkit end
 
 import co.aikar.timings.SpigotTimings; // Spigot
+import pw.alphheim.api.events.EffectRemovalNotification;
 
 // PaperSpigot start
 import org.bukkit.Bukkit;
+import org.bukkit.potion.PotionEffect;
+import org.bukkit.potion.PotionEffectType;
 import org.spigotmc.event.entity.EntityDismountEvent;
 // PaperSpigot end
 
@@ -481,6 +484,7 @@ public abstract class EntityLiving extends Entity {
                 if (!this.world.isClientSide) {
                     iterator.remove();
                     this.b(mobeffect);
+                    new EffectRemovalNotification(this.getBukkitEntity(), new PotionEffect(PotionEffectType.getById(mobeffect.getEffectId()), mobeffect.getDuration(), mobeffect.getAmplifier(), mobeffect.isAmbient(), mobeffect.isShowParticles())).callEvent(); // Alphheim - event for notifying effect removal
                 }
             } else if (mobeffect.getDuration() % 600 == 0) {
                 this.a(mobeffect, false);
-- 
2.17.0

