From 80168d57d909b4facba866671a8c8f3395725981 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 16 Jan 2018 04:53:22 +0000
Subject: [PATCH] Allow overriding enchantment limitations

---
 .../net/minecraft/server/CommandEnchant.java  | 22 ++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/src/main/java/net/minecraft/server/CommandEnchant.java b/src/main/java/net/minecraft/server/CommandEnchant.java
index ea6aa228..384411b8 100644
--- a/src/main/java/net/minecraft/server/CommandEnchant.java
+++ b/src/main/java/net/minecraft/server/CommandEnchant.java
@@ -26,6 +26,12 @@ public class CommandEnchant extends CommandAbstract {
             throw new ExceptionUsage("commands.enchant.usage", new Object[0]);
         } else {
             EntityLiving entityliving = (EntityLiving) a(minecraftserver, icommandlistener, astring[0], EntityLiving.class);
+            // Alphheim start - bypass enchant limitations
+            boolean bypassLimit = false;
+            if (astring.length == 4) {
+                bypassLimit = Boolean.parseBoolean(astring[3]);
+            }
+            // Alphheim end - bypass enchant limitations
 
             icommandlistener.a(CommandObjectiveExecutor.EnumCommandResult.AFFECTED_ITEMS, 0);
 
@@ -45,14 +51,17 @@ public class CommandEnchant extends CommandAbstract {
 
                 if (itemstack.isEmpty()) {
                     throw new CommandException("commands.enchant.noItem", new Object[0]);
-                } else if (!enchantment.canEnchant(itemstack)) {
+                } else if (!bypassLimit && !enchantment.canEnchant(itemstack)) { // Alphheim - bypass enchant limitations
                     throw new CommandException("commands.enchant.cantEnchant", new Object[0]);
                 } else {
+
+                    // Alphheim start - allow bypassing enchantment limits
                     if (astring.length >= 3) {
-                        i = a(astring[2], enchantment.getStartLevel(), enchantment.getMaxLevel());
+                        i = bypassLimit ? a(astring[2]) : a(astring[2], enchantment.getStartLevel(), enchantment.getMaxLevel());
                     }
 
-                    if (itemstack.hasTag()) {
+                    if (itemstack.hasTag() && !bypassLimit) {
+                    // Alphheim end - allow bypassing enchantment limits
                         NBTTagList nbttaglist = itemstack.getEnchantments();
 
                         for (int j = 0; j < nbttaglist.size(); ++j) {
@@ -83,4 +92,11 @@ public class CommandEnchant extends CommandAbstract {
     public boolean isListStart(String[] astring, int i) {
         return i == 0;
     }
+
+    // Alphheim start - fix decompile error
+    @Override
+    public int compareTo(ICommand o) {
+        return a(o);
+    }
+    // Alphheim end
 }
-- 
2.17.1

