From 0453c9d88708e83ceb701dd31c1bab9b4cc3941c Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Thu, 27 Sep 2018 04:22:36 +0100
Subject: [PATCH] implement equals/hashcode on Chunk

---
 .../java/org/bukkit/craftbukkit/CraftChunk.java | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index 011e40458e..3c2a5ca581 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -338,4 +338,21 @@ public class CraftChunk implements Chunk {
     static {
         Arrays.fill(emptyLight, (byte) 0xFF);
     }
+
+    // Alphheim start
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        CraftChunk that = (CraftChunk) o;
+        return x == that.x &&
+                z == that.z &&
+                worldServer.equals(that.worldServer);
+    }
+
+    @Override
+    public int hashCode() {
+        return java.util.Objects.hash(worldServer, x, z);
+    }
+    // Alphheim end
 }
-- 
2.21.0

