From 7eb9a6758ea74b734d3e867a8c90d49f61a4a0be Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 16 Jan 2018 04:53:22 +0000
Subject: [PATCH] Allow overriding enchantment limitations

---
 .../net/minecraft/server/CommandEnchant.java  | 25 ++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/src/main/java/net/minecraft/server/CommandEnchant.java b/src/main/java/net/minecraft/server/CommandEnchant.java
index 500a6a94..81c33ce4 100644
--- a/src/main/java/net/minecraft/server/CommandEnchant.java
+++ b/src/main/java/net/minecraft/server/CommandEnchant.java
@@ -23,6 +23,15 @@ public class CommandEnchant extends CommandAbstract {
         if (astring.length < 2) {
             throw new ExceptionUsage("commands.enchant.usage", new Object[0]);
         } else {
+
+            // Alphheim start - bypass enchant limitations
+            boolean bypassLimit = false;
+            if (astring.length == 4) {
+                bypassLimit = Boolean.parseBoolean(astring[3]);
+            }
+
+            // Alphheim end - bypass enchant limitations
+
             EntityPlayer entityplayer = a(icommandlistener, astring[0]);
 
             icommandlistener.a(CommandObjectiveExecutor.EnumCommandResult.AFFECTED_ITEMS, 0);
@@ -51,14 +60,17 @@ public class CommandEnchant extends CommandAbstract {
 
                 if (enchantment1 == null) {
                     throw new ExceptionInvalidNumber("commands.enchant.notFound", new Object[] { Integer.valueOf(i)});
-                } else if (!enchantment1.canEnchant(itemstack)) {
+                } else if (!bypassLimit && !enchantment1.canEnchant(itemstack)) { // Alphheim - bypass enchant limitations
                     throw new CommandException("commands.enchant.cantEnchant", new Object[0]);
                 } else {
+
+                    // Alphheim start - allow bypassing enchantment limits
                     if (astring.length >= 3) {
-                        j = a(astring[2], enchantment1.getStartLevel(), enchantment1.getMaxLevel());
+                        j = bypassLimit ? a(astring[2]) : a(astring[2], enchantment1.getStartLevel(), enchantment1.getMaxLevel());
                     }
 
-                    if (itemstack.hasTag()) {
+                    if (itemstack.hasTag() && !bypassLimit) {
+                    // Alphheim end - allow bypassing enchantment limits
                         NBTTagList nbttaglist = itemstack.getEnchantments();
 
                         if (nbttaglist != null) {
@@ -95,4 +107,11 @@ public class CommandEnchant extends CommandAbstract {
     public boolean isListStart(String[] astring, int i) {
         return i == 0;
     }
+
+    // Alphheim start - fix decompile error
+    @Override
+    public int compareTo(ICommand o) {
+        return a(o);
+    }
+    // Alphheim end
 }
-- 
2.17.1

